alphastats/DataSet_Preprocess.py

```diff
@@ -215,64 +251,68 @@ def _normalization(self, method: str):
-                imputation=preprocessing_mode[1],
-            )
-
-            res = func(**params_for_func)
-            results_list.append(res)
-
-            print("\t")
-
-        return results_list
+    # TODO this needs to be reimplemented
+    # @ignore_warning(RuntimeWarning)
+    # def _compare_preprocessing_modes(self, func, params_for_func) -> list:
+    #     dataset = self
+    #
+    #     preprocessing_modes = list(
+    #         itertools.product(self.normalization_methods, self.imputation_methods)
+    #     )
+    #
+    #     results_list = []
+    #
+    #     del params_for_func["compare_preprocessing_modes"]
+    #     params_for_func["dataset"] = params_for_func.pop("self")
+    #
+    #     # TODO: make this progress transparent in GUI
+    #     for preprocessing_mode in tqdm(preprocessing_modes):
+    #         # reset preprocessing
+    #         dataset.reset_preprocessing()
+    #         print(
+    #             f"Normalization {preprocessing_mode[0]}, Imputation {str(preprocessing_mode[1])}"
+    #         )
+    #         dataset.mat.replace([np.inf, -np.inf], np.nan, inplace=True)
+    #
+    #         dataset.preprocess(
+    #             subset=True,
+    #             normalization=preprocessing_mode[0],
+    #             imputation=preprocessing_mode[1],
+    #         )
+    #
+    #         res = func(**params_for_func)
+    #         results_list.append(res)
+    #
+    #         print("\t")
+    #
+    #     return results_list
 
     def _log2_transform(self):
         self.mat = np.log2(self.mat + 0.1)
         self.preprocessing_info.update({"Log2-transformed": True})
         print("Data has been log2-transformed.")
 
-    def batch_correction(self, batch: str):
+    def batch_correction(self, batch: str) -> Tuple[pd.DataFrame, pd.DataFrame]:
         """Correct for technical bias/batch effects
-        Behdenna A, Haziza J, Azencot CA and Nordor A. (2020) pyComBat, a Python tool for batch effects correction in high-throughput molecular data using empirical Bayes methods. bioRxiv doi: 10.1101/2020.03.17.995431
+
         Args:
             batch (str): column name in the metadata describing the different batches
+
+        # TODO should the employed methods (and citations) be made transparent in the UI?
+        Behdenna A, Haziza J, Azencot CA and Nordor A. (2020) pyComBat,
+        a Python tool for batch effects correction in high-throughput molecular
+        data using empirical Bayes methods. bioRxiv doi: 10.1101/2020.03.17.995431
         """
         from combat.pycombat import pycombat
 
         data = self.mat.transpose()
         series_of_batches = self.metadata.set_index(self.sample).reindex(
             data.columns.to_list()
         )[batch]
-        self.mat = pycombat(data=data, batch=series_of_batches).transpose()
+
+        batch_corrected_data = pycombat(data=data, batch=series_of_batches).transpose()
+
+        return batch_corrected_data, self.metadata
```

Comment:
self. metadata does not actually change, so it doesn't need to be returned.

Reply:
it does in `_remove_samples()` (this is the bug that I think this fixes: on `reset_processing`, metadata is not reset)

Reply:
update: you're right, in `batch_processing`, this is not touched .. thanks for spotting!